### PR TITLE
Extract diffing functionality as a new semantic-diff package

### PR DESCRIPTION
**Pull request by tclem**
In service of https://github.com/github/semantic/issues/438, this extracts semantic's diffing capabilities into a dedicated package. This makes the API boundaries a bit more obvious and creates a path for restructuring the data types so that diffing works properly on precise ASTs.

### CODE REVIEW

File: bench/Tagging.hs [hunk 1]
```diff
@@ -10,7 +10,7 @@ import           Control.Exception (throwIO)
 import           Control.Monad
 import           Data.Blob
 import           Data.Foldable
-import           Data.Language (LanguageMode (..), PerLanguageModes (..))
+import           Data.Language (PerLanguageModes (..), aLaCarteLanguageModes, preciseLanguageModes)
```
Comment (robrix) [inline bench/Tagging.hs]: Good call.

File: script/ghci-flags [hunk 2]
```diff
@@ -37,6 +37,7 @@ function flags {
   echo "-isemantic-analysis/src"
   echo "-isemantic-ast/src"
   echo "-isemantic-core/src"
+  echo "-isemantic-diff/src"
```
Comment (robrix) [inline script/ghci-flags]: ❤️

I’ve been meaning to add a note to the `cabal.project{,.ci}` files to remember to do this, so thank you for remembering anyway 😊

File: script/ghci-flags-dependencies [hunk 3]
```diff
@@ -11,6 +11,7 @@ echo "semantic.cabal"
 echo "semantic-analysis/semantic-analysis.cabal"
 echo "semantic-ast/semantic-ast.cabal"
 echo "semantic-core/semantic-core.cabal"
+echo "semantic-analysis/semantic-diff.cabal"
```
Comment (robrix) [inline script/ghci-flags-dependencies]:
```suggestion
echo "semantic-diff/semantic-diff.cabal"
```

File: semantic-diff/semantic-diff.cabal [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+                   , Data.Edit
+                   , Data.Functor.Classes.Generic
+                   , Data.JSON.Fields
+                   , Data.Term
+                   , Diffing.Algorithm
+                   , Diffing.Algorithm.RWS
+                   , Diffing.Algorithm.RWS.FeatureVector
+                   , Diffing.Algorithm.SES
+                   , Diffing.Interpreter
+                   , Prologue
```
Comment (robrix) [inline semantic-diff/semantic-diff.cabal]: 1. The `exposed-modules` field accepts formatting without commas:

    ```cabal
    exposed-modules:
      Data.Diff
      Data.Algebra
      …
    ```

    I find this style supports nicer diffing; could you apply it here please? 🙏

2. Could you please alphabetize the list of modules? It makes finding things and updating the list simpler, tho it is unfortunately rather manual at the moment (absent something like `cabal-fmt`).

3. Instead of exporting `Prologue`, could you please either a) remove it outright from this package, or b) move it to an `other-modules` field?

File: semantic-diff/semantic-diff.cabal [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+                   , mersenne-random-pure64 ^>= 0.2.2.0
+                   , array ^>= 0.5.3.0
+                   , generic-monoid ^>= 0.1.0.0
+                   , kdt ^>= 0.2.4
+                   , mtl ^>= 2.2.2
+                   , hashable >= 1.2.7 && < 1.4
+                   , fused-effects ^>= 1
+                   , semilattices ^>= 0.0.0.3
+                   , bytestring ^>= 0.10.8.2
+                   , bifunctors ^>= 5.5
```
Comment (robrix) [inline semantic-diff/semantic-diff.cabal]: `build-depends` sadly doesn’t accept comma-free formatting, but formatting the actual dependencies to start on the next line & alphabetizing makes it easier to diff & update this, too:

```cabal
build-depends:
    aeson^>= 1.4.2.0
  , base >= 4.13 && < 5
  …
```

File: semantic-diff/src/Data/Algebra.hs [hunk 6]
```diff
@@ -0,0 +1,48 @@
+{-# LANGUAGE DeriveGeneric, DeriveTraversable, RankNTypes #-}
+module Data.Algebra
```
Comment (robrix) [inline semantic-diff/src/Data/Algebra.hs]: I think eventually we’re going to either remove this or replace it with something better-suited to precise ASTs in `semantic-ast`, but this is a good home for it for the time being 👍

File: semantic-tsx/semantic-tsx.cabal [hunk 7]
```diff
@@ -27,6 +27,7 @@ common haskell
                      , semantic-core ^>= 0.0
                      , semantic-source ^>= 0.0.1
                      , semantic-tags ^>= 0.0
+                     , semantic-diff
```
Comment (robrix) [inline semantic-tsx/semantic-tsx.cabal]: I don’t know if we should be specifying version bounds here or not, but the difference between this and the preceding lines gives me pause. We need one for `semantic-source`, presumably, but maybe we shouldn’t specify them for _any_ of the local packages since we aren’t getting them from hackage? It only makes more work for us in keeping these up to date… if we ever update them. Which I suppose we won’t until we start moving them to hackage.

So I guess that makes this academic.

File: semantic-tsx/src/Language/TSX/Tags.hs [hunk 8]
```diff
@@ -100,6 +100,19 @@ instance ToTags Tsx.Class where
...
+    { ann = loc@Loc { byteRange }
+    , name
+    } = match name
+    where
+      match expr = case expr of
+        Prj Tsx.Identifier { text } -> yield text
+        -- TODO: Handle NestedIdentifiers and Strings
+        -- Prj Tsx.NestedIdentifier { extraChildren } -> match
+        _ -> gtags t
+      yield text = yieldTag text Module loc byteRange >> gtags t
```
Comment (robrix) [inline semantic-tsx/src/Language/TSX/Tags.hs]: Does this change belong to another PR? (No worries if so, I just want to be sure we don’t merge something somewhere it shouldn’t be prematurely/by mistake.)

File: src/Serializing/SExpression/Precise.hs [hunk 9]
```diff
@@ -78,5 +78,5 @@ instance ToSExpression t => GToSExpression (Rec1 t) where
 
 instance (Foldable f, GToSExpression g) => GToSExpression (f :.: g) where
   gtoSExpression (Comp1 fs) n
-    | null fs   = [nl n <> pad n <> "[]"]
-    | otherwise = nl n <> pad n <> "[" : foldMap gtoSExpression fs (n + 1) <> ["]"]
+    | null fs   = mempty
+    | otherwise = foldMap gtoSExpression fs n
```
Comment (robrix) [inline src/Serializing/SExpression/Precise.hs]: Is the intention here to mimic the existing s-expression output’s style for tests? I’d actually prefer if we _improved_ on that style, in general, since it doesn’t e.g. make it clear when there are multiple subterms in a single field, or when there is a field with no subterms. (I seem to recall we had an issue about this at one point, but I can’t find it just now.) Thoughts?

**Comment by patrickt** [comment]
@tclem How do we want to proceed with this? Do we want to disable building it for the time being (i.e. remove it from `cabal.project`), given that #438 seems to indicate that we want to disable it while we’re moving the alacarte ASTs out of this repo?

**Comment by tclem** [comment]
I think it really depends on if we want to yank diffing support out of semantic entirely right now or if we want to maintain the basic algorithms... I don't think it's worth merging or working on this anymore until we can make that decision first.

**Comment by patrickt** [comment]
@tclem The whole shape of pretty much everything in the diffing stuff will change, right, as the shapes of the types in question are going from `* -> *` to `(* -> *) -> * -> *`? Are we gonna be able to preserve anything in the basic algorithms themselves?

**Comment by tclem** [comment]
Closing in favor of https://github.com/github/semantic/pull/570 as the team consensus is to remove instead of extract, knowing that the shape of these algorithms will be different when/if we add them back in.
